elastic-coders/serverless-webpack
# Support individual packaging with multi-compile

## What did you implement:

Closes #181 
Closes #120 
Closes #173 
Closes #117 
Closes #116 
Relates to #60 

## How did you implement it:

If `package.individually: true` is set in the service, the individual entry points (functions) are now compiled by using WebPack's multi-compile feature. That means, that each function is optimized separately as well as the externals for every function. The node_modules for each function are then packaged individually, so that every resulting function zip contains only the needed modules and the individually optimized function handler.

Additionally the plugin does not change `serverless.config.servicePath` anymore, so that other plugins and the framework itself behave as expected and work during any lifecycle event and Serverless does not interfere with the packaging of the optimized node modules anymore.

### Further changes
* Added ESLint as preparation for 3.0.0, so that the implementation is conformant.
* Added unit tests for the new functionality (as well as for old ones where tests where missing).
* Added verbose output with `--verbose`

## How can we verify it:

Use a simple service that defines multiple functions, where each function uses a different set of dependencies. Use a webpack config that uses `lib.entries` for entry and set the package wide individual packaging in the service.

Now run `serverless package` and check the generated function zip files in the `.serverless` folder.

I checked it with one of our production projects, including tests with `serverless invoke local`. Found no issues. The only thing I have to verify is, that deployments without individual packaging still work.

## Todos:

- Fix linting errors (not yet implemented)
- Make sure code coverage hasn't dropped (not yet implemented)

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
